### PR TITLE
Fix+upgrade mysql-deps in both images

### PR DESCRIPTION
**Pull request by benbender**
+ small typo-fix ;)

### CODE REVIEW

**Comment by maxking** [comment]
Thanks for you contribution! I like the support for mariadb, I thought it was fully compatible with mysql libraries and didn't need any extra libraries.

Is there a reason for moving to mysqlclient instead of pymysql though? Both seem to be under the same organization on Github and maintained by same developer. Any advantages of mysqlclient? Looks like it is a fork of the old mysql library.

Either way, the best way would be to first propose this change in Mailman Core to see if the test suite is fully compatible this library and then we can change it here.

**Comment by benbender** [comment]
I thought pymysql is not compatible with python3 - but that seems to be not the case. I think I was confused by mysqlclient/pymysql/mysqlDb. Anyway I just played around and don't have the time available to bring this any further. At least, this changeset makes the images really compatible with mysql. ATM master can't connect or be build with mysql in mind - if you want to push it upstream, feel free! Thanks for your work!

**Comment by maxking** [comment]
Core actually does work with MySQL using pymysql so it must be something missing in the images which makes it in-compatible with MySQL. I will go ahead and fix that part.

Thanks!

**Comment by benbender** [comment]
I'm really sorry, but the images don't work with mysql. Clean checkout + build with this docker-compose.yml:

```yaml
version: '2'

services:
  nginx:
    image: nginx
    container_name: nginx
    hostname: nginx
    volumes:
      - ./mailman.conf:/etc/nginx/conf.d/mailman.conf
      - ./data/web/static:/mailman/static
    ports:
      - 8080:80
    command: nginx -g 'daemon off;'
    networks:
      mailman:
        ipv4_address: 172.19.199.2

  mailman-core:
    image: maxking/mailman-core:0.1
    container_name: mailman-core
    hostname: mailman-core
    volumes:
      - ./data/core:/opt/mailman/
    links:
      - database:database
    depends_on:
      - database
    environment:
      - DATABASE_TYPE=mysql
      - DATABASE_URL=mysql://mailman:foobar@database/mailman?charset=utf8&use_unicode=1
      - DATABASE_CLASS=mailman.database.mysql.MySQLDatabase
      - HYPERKITTY_API_KEY=someapikey
      - HYPERKITTY_URL=http://mailman-web:8000/hyperkitty
    networks:
      mailman:
        ipv4_address: 172.19.199.2

  mailman-web:
    image: maxking/mailman-web:0.1
    container_name: mailman-web
    hostname: mailman-web
    depends_on:
      - database
    links:
      - mailman-core:mailman-core
      - database:database
    volumes:
      - ./data/web:/opt/mailman-web-data
    expose:
      - 8000
      - 8080
    environment:
      - DATABASE_TYPE=mysql
      - DATABASE_URL=mysql://mailman:foobar@database/mailman?charset=utf8&use_unicode=1
      - HYPERKITTY_API_KEY=someapikey
      - SECRET_KEY=somesecretkey
    networks:
      mailman:
        ipv4_address: 172.19.199.3

  database:
    image: mariadb:latest
    container_name: database
    hostname: database
    restart: unless-stopped
    expose:
      - 3306
    command: mysqld --max_allowed_packet=192M --max-connections=1500 --innodb-strict-mode=0 --skip-host-cache --skip-name-resolve --log-warnings=0
    environment:
      - MYSQL_ROOT_PASSWORD=foobar
      - MYSQL_USER=mailman
      - MYSQL_PASSWORD=foobar
      - MYSQL_DATABASE=mailman
    networks:
      mailman:
        ipv4_address: 172.19.199.4

networks:
   mailman:
     driver: bridge
     ipam:
       driver: default
       config:
       -
         subnet: 172.19.199.0/24
````

Leads to these errors on `docker-compose up`:

````
$ docker-compose logs mailman-core
Attaching to mailman-core
mailman-core    | Traceback (most recent call last):
mailman-core    |   File "/usr/local/bin/mailman", line 11, in <module>
mailman-core    |     sys.exit(main())
mailman-core    |   File "/usr/local/lib/python3.6/site-packages/mailman/bin/mailman.py", line 97, in main
mailman-core    |     initialize(config_path)
mailman-core    |   File "/usr/local/lib/python3.6/site-packages/mailman/core/initialize.py", line 185, in initialize
mailman-core    |     initialize_2(propagate_logs=propagate_logs)
mailman-core    |   File "/usr/local/lib/python3.6/site-packages/mailman/core/initialize.py", line 156, in initialize_2
mailman-core    |     config.db = getUtility(IDatabaseFactory, utility_name).create()
mailman-core    |   File "/usr/local/lib/python3.6/site-packages/mailman/database/factory.py", line 54, in create
mailman-core    |     database.initialize()
mailman-core    |   File "/usr/local/lib/python3.6/site-packages/mailman/database/base.py", line 106, in initialize
mailman-core    |     self.engine = create_engine(url, isolation_level='READ UNCOMMITTED')
mailman-core    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/__init__.py", line 387, in create_engine
mailman-core    |     return strategy.create(*args, **kwargs)
mailman-core    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/strategies.py", line 80, in create
mailman-core    |     dbapi = dialect_cls.dbapi(**dbapi_args)
mailman-core    |   File "/usr/local/lib/python3.6/site-packages/sqlalchemy/dialects/mysql/mysqldb.py", line 110, in dbapi
mailman-core    |     return __import__('MySQLdb')
mailman-core    | ModuleNotFoundError: No module named 'MySQLdb'
````
And same for web:

```
$ docker-compose logs mailman-web
Attaching to mailman-web
mailman-web     | settings_local.py not found, it is highly recommended that you provide one
mailman-web     | Using default configuration to run.
mailman-web     | Traceback (most recent call last):
mailman-web     |   File "manage.py", line 10, in <module>
mailman-web     |     execute_from_command_line(sys.argv)
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
mailman-web     |     utility.execute()
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 341, in execute
mailman-web     |     django.setup()
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/__init__.py", line 27, in setup
mailman-web     |     apps.populate(settings.INSTALLED_APPS)
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate
mailman-web     |     app_config.import_models(all_models)
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/apps/config.py", line 199, in import_models
mailman-web     |     self.models_module = import_module(models_module_name)
mailman-web     |   File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
mailman-web     |     __import__(name)
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/hyperkitty/models/__init__.py", line 27, in <module>
mailman-web     |     from .category import ThreadCategory
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/hyperkitty/models/category.py", line 30, in <module>
mailman-web     |     class ThreadCategory(models.Model):
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/db/models/base.py", line 119, in __new__
mailman-web     |     new_class.add_to_class('_meta', Options(meta, app_label))
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/db/models/base.py", line 316, in add_to_class
mailman-web     |     value.contribute_to_class(cls, name)
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/db/models/options.py", line 214, in contribute_to_class
mailman-web     |     self.db_table = truncate_name(self.db_table, connection.ops.max_name_length())
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/db/__init__.py", line 36, in __getattr__
mailman-web     |     return getattr(connections[DEFAULT_DB_ALIAS], item)
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/db/utils.py", line 211, in __getitem__
mailman-web     |     backend = load_backend(db['ENGINE'])
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/db/utils.py", line 115, in load_backend
mailman-web     |     return import_module('%s.base' % backend_name)
mailman-web     |   File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
mailman-web     |     __import__(name)
mailman-web     |   File "/usr/local/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 28, in <module>
mailman-web     |     raise ImproperlyConfigured("Error loading MySQLdb module: %s" % e)
mailman-web     | django.core.exceptions.ImproperlyConfigured: Error loading MySQLdb module: No module named MySQLdb
```

**Comment by maxking** [comment]
Ah! I am sorry about it but it does makes sense. I have never tested Django with MySQL, so it not working is probably my mistake (sorry!), I will add the relevant python library and also the tests for it.

The Core however should work with pymysql, the `DATABASE_URL=mysql+pymysql://myuser:mypassword@mymysqlhost/mailman?charset=utf8&use_unicode=1` should work.

Without `pymysql` in the above URL, sqlalchemy tries to use the default mysql library which doesn't work because it doesn't have Python3 support (also because it is missing in the images).

**Comment by maxking** [comment]
The DATABASE_URL mentioned above is a generic one and needs replacing of values to work with containers.

**Comment by benbender** [comment]
I tried it with the url-scheme from the documentation, but unfortunatly this gets you an out of index-exception. The installed Django doesn't seem to know the scheme.

**Comment by maxking** [comment]
Yes, that scheme is only  meant for Mailman Core, Django is probably missing a library.

**Comment by benbender** [comment]
I'll give it a shot on core.

**Comment by benbender** [comment]
Meanwhile: It works for core, but not for web. I'll dig a bit deeper...

**Comment by maxking** [comment]
It looks like `mysqlclient` is indeed the recommended library for MySQL and Django.

https://docs.djangoproject.com/en/1.11/ref/databases/#mysql-db-api-drivers

If you can remove the changes to core/Dockerfile, I can merge this. Thanks for this!

**Comment by maxking** [comment]
Also if you can add a version requirement `>=1.3.3` for mysqlclient it would be great, it looks like this is the min required for Django.

https://docs.djangoproject.com/en/1.11/ref/databases/#id9

**Comment by benbender** [comment]
done

File: core/Dockerfile [hunk 1]
```diff
@@ -7,10 +7,9 @@ COPY docker-entrypoint.sh /usr/local/bin/
 
 #Install all required packages, add user for executing mailman and set execution rights for startup script
 RUN apk update \
-    && apk add --virtual build-deps gcc python3-dev musl-dev \
-    && apk add postgresql-dev bash su-exec postgresql-client \
-    && pip install psycopg2 mailman==3.1.0 mailman-hyperkitty==1.1.0 \
-	&& pip install pymysql \
+    && apk add --virtual build-deps gcc python3-dev musl-dev postgresql-dev \
+    && apk add bash su-exec postgresql-client  \
+    && pip install mailman mailman-hyperkitty psycopg2 pymysql \
```
Comment (maxking) [inline core/Dockerfile]: Any reason for removing the pinned versions?

I try to keep the versions of Mailman components deterministic in the images, any new release is added in newer versions of images.

File: web/Dockerfile [hunk 2]
```diff
@@ -20,8 +20,8 @@ RUN set -ex \
 		uwsgi \
 		psycopg2 \
 		dj-database-url \
-		pymysql \
-	&& pip install -U django==1.10 \
```
Comment (maxking) [inline web/Dockerfile]: I have seen before that if django is specified as a dependency for any of the components mentioned above, the version directive is ignored assuming that dependency is already satisfied.
To force installation of a specific version, I had to run it as a separate command here.

File: web/Dockerfile [hunk 3]
```diff
@@ -10,7 +10,7 @@ COPY docker-entrypoint.sh /usr/local/bin/
 # Install packages and dependencies for postorius and hyperkitty
 # Add user for executing apps, change ownership for uwsgi+django files and set execution rights for management script
 RUN set -ex \
-	&& apk add --no-cache --virtual .build-deps gcc libc-dev linux-headers postgresql-dev \
+	&& apk add --no-cache --virtual .build-deps gcc libc-dev linux-headers postgresql-dev mariadb-dev \
```
Comment (maxking) [inline web/Dockerfile]: Is `mysql-dev` not required here for mysql? 
I understand that you have been testing with mariadb, would this library be enough to work with mysql as well?

**Comment by lonewolfnight** [comment]
Changes as proposed continue to throw errors for mysql 

Copying settings_local.py ...
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 341, in execute
    django.setup()
  File "/usr/local/lib/python2.7/site-packages/django/__init__.py", line 27, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python2.7/site-packages/django/apps/registry.py", line 108, in populate
    app_config.import_models(all_models)
  File "/usr/local/lib/python2.7/site-packages/django/apps/config.py", line 199, in import_models
    self.models_module = import_module(models_module_name)
  File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/usr/local/lib/python2.7/site-packages/hyperkitty/models/__init__.py", line 27, in <module>
    from .category import ThreadCategory
  File "/usr/local/lib/python2.7/site-packages/hyperkitty/models/category.py", line 30, in <module>
    class ThreadCategory(models.Model):
  File "/usr/local/lib/python2.7/site-packages/django/db/models/base.py", line 119, in __new__
    new_class.add_to_class('_meta', Options(meta, app_label))
  File "/usr/local/lib/python2.7/site-packages/django/db/models/base.py", line 316, in add_to_class
    value.contribute_to_class(cls, name)
  File "/usr/local/lib/python2.7/site-packages/django/db/models/options.py", line 214, in contribute_to_class
    self.db_table = truncate_name(self.db_table, connection.ops.max_name_length())
  File "/usr/local/lib/python2.7/site-packages/django/db/__init__.py", line 36, in __getattr__
    return getattr(connections[DEFAULT_DB_ALIAS], item)
  File "/usr/local/lib/python2.7/site-packages/django/db/utils.py", line 211, in __getitem__
    backend = load_backend(db['ENGINE'])
  File "/usr/local/lib/python2.7/site-packages/django/db/utils.py", line 115, in load_backend
    return import_module('%s.base' % backend_name)
  File "/usr/local/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/usr/local/lib/python2.7/site-packages/django/db/backends/mysql/base.py", line 28, in <module>
    raise ImproperlyConfigured("Error loading MySQLdb module: %s" % e)
django.core.exceptions.ImproperlyConfigured: Error loading MySQLdb module: No module named MySQLdb

**Comment by maxking** [comment]
Something very weird is happening when I try this, looks like that the migrations succeed but the access after that somehow doesn't work. There is something really weird going on.

I'd have to spend more time on it to debug the exact issue here.

**Comment by lonewolfnight** [comment]
We've tracked this down some clues. The failure of module is associated to usr/local/lib/python2.7/site-packages/django/db/backends/mysql/base.py. This is not the fix nor the problem but we worked a overlay to append the package directly. This is obviously not the solution in any way but may provide debug help. 

import pymysql
pymysql.install_as_MySQLdb()

**Comment by maxking** [comment]
pymysql isn't listed as a supported library for Django, so I don't expect it to work. it *might* work with `pymysql.install_as_MySQLdb()` as you mentioned, but I would prefer to use `mysqlclient` as recommended in Django docs.

In my testing, I was able to get to migrations, and they were applied to the MySQL server I was running. So, I know atleast that mysqlclient works with Django, but then Django crashed the first time it tried to "access" the Database after it started running.

I'd first try to run it locally, without any containers to see if mysqlclient works nicely with Django and MySQL. After that, it would help to pinpoint if the issue is with library or our docker setup.

**Comment by maxking** [comment]
Thanks for your help with this, I manually merged your changes in #140 

The latest images should work wit MySQL (they are tested with mysql now).

**Pull request closed without merging**